[web] Fix: dns_diagnostics.php SPF DNS lookup loop

ualnwe opened the pull request:
### Description

### The problem:
When using dns diagnostics to check SPF records, a possible loop can occur.
If the domain has an SPF record pointing to another domain that has an SPF record referencing the original via `include:`.
This causes an infinite stall.

### Example: 
- call dns_diagnostics.php?domain=example.com
- example.com has an SPF record with `include:mail.example.com`
- mail.example.com has an SPF record with `include:example.com`

### The fix:
I appended `, $original_domain = ""` to the `get_spf_allowed_hosts(..` function in `data/web/inc/spf.inc.php` and also set every call within `spf.inc.php` to pass `$check_domain` to this new argument. Then by adding an additional check on line 50 for this original domain ` && $original_domain != $domain` it correctly stops after one loop as we're back on the original domain.

Closes #4715

ualnwe committed:
Add check for looping SPF record

Keep track of the original domain queried to check we aren't looping between SPF include records.

ualnwe committed:
fix spaces -> tabs

mkuron reviewed (changes requested):
Your change only checks for a loop back to the original domain -- before it can be merged, it needs to be extended to detect _all_ loops, including those that loop back to intermediate domains. Also, your pull request contains some unrelated changes that need to be removed.

ualnwe committed:
Use global var to track all loop domains

Changed to use a global array that tracks domains checked, stopping us from checking the same domain more than once

ualnwe committed:
Use the actual domain getting checked

ualnwe commented:
This should now keep track of checked domains and detect a loop using a global array.
There's probably a better way to do this, but this works in my testing :sweat_smile: 
Apologies for the unrelated changes, and I may have selected the wrong branch.
Thank you DerLinkman. Hopefully these both are now resolved.

ualnwe committed:
fix correct SPF + handle looping redirects

mkuron commented on data/web/inc/spf.inc.php:
```diff
@@ -1,6 +1,8 @@
-function get_spf_allowed_hosts($check_domain, $expand_ipv6 = false, $original_domain = "") {
+$checked_domains = array();
+function get_spf_allowed_hosts($check_domain, $expand_ipv6 = false) {
+	global $checked_domains;
```
Please don't use global variables. This fill break the loop detection when multiple SPF checks are performed on the same page.

ualnwe replied:
Sorry, for some reason I didn't think about just passing the array back into the function.
I've changed it to pass in a new optional argument, the array containing all previously checked include or redirect SPF domain records.

ualnwe committed:
Merge branch 'mailcow:staging' into staging

ualnwe committed:
Replace global with optional arg

mkuron commented:
Thanks, that's much better. Sorry for keeping requesting changes, but I think it might make sense to display a warning on the DNS diagnostics page if an SPF loop is detected.

I checked RFC 7208 and it actually says this:
> Some mechanisms and modifiers (collectively, "terms") cause DNS
   queries at the time of evaluation, and some do not.  The following
   terms cause DNS queries: the "include", "a", "mx", "ptr", and
   "exists" mechanisms, and the "redirect" modifier.  SPF
   implementations MUST limit the total number of those terms to 10
   during SPF evaluation, to avoid unreasonable load on the DNS.  If
   this limit is exceeded, the implementation MUST return "permerror".

So simply ignoring looping redirects is not RFC compliant. Instead, we should actually treat this as an error condition. Depending on the other server's SPF implementation, a permerror may either be considered an SPF pass or an SPF fail, so you won't be able to reliably deliver emails if your SPF has a loop.

DerLinkman commented:
Is still a thing?

mkuron commented:
Yes, I'd like to get that merged, but @ualnwe never addressed my latest comment. Sorry that it took multiple rounds of review (I hope you didn't lose interest because of that, @ualnwe), I should have looked at the RFC first.

ualnwe commented:
Apologies for the lack of updates, I've been rather busy with other things.
I'd still very much like to fix this, and to the correct standard as mentioned re: being RFC compliant

milkmaker commented:
This pull request has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs.
